### ISSUE TITLE
SearchController.closeView failure on PWA after upgrade to 3.16.0

### ISSUE THREAD

**Issue by lyskouski** [flutter/flutter#138880]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Install Web [app](https://tercad.com/app/finance/index.html) [v5.1.4](https://github.com/lyskouski/app-finance/releases/tag/v5.1.4) as PWA (progressive web app) on mobile device
2. Open it
3. Use language selector to switch locale to Chinese

### Expected results

Selector is closed.

### Actual results

Fatal error: Null check operator used on a null value

NOTE: Reproduced only for PWA. Works for all other devices.
NOTE: Not relevant to a language selector, happens with any SearchAnchor (check attachments)

### Code sample

<details><summary>Code sample</summary>

```dart

class ListSelectorItem {
  final String id;
  final String name;

  bool match(String search) => search.isPartOf(name, Locale(AppLocale.code));

  bool equal(val) => id == val;

  @override
  toString() => name;

  ListSelectorItem({required this.id, required this.name});
}


class ListSelector<K extends ListSelectorItem> extends AbstractSelector {
  final List<K> options;
  final Function setState;
  final String? hintText;
  final String? tooltip;
  final TextStyle? hintStyle;
  final Color? hintColor;
  final bool withLabel;

  const ListSelector({
    super.key,
    required this.options,
    required this.setState,
    this.hintText,
    this.tooltip,
    this.hintStyle,
    this.hintColor,
    super.value,
    this.withLabel = false,
  });

  @override
  ListSelectorState createState() => ListSelectorState();
}

class ListSelectorState<T extends ListSelector, K extends ListSelectorItem> extends AbstractSelectorState<T> {
  Widget selectorBuilder(BuildContext context, K item) => Text(item.toString());

  Widget itemBuilder(BuildContext context, K item) => selectorBuilder(context, item);

  void onChange(K value) {
    textController.closeView(null); // [!Error]
    widget.setState(value.id);
  }

  @override
  void onTap(BuildContext? context) {
    focusController.onFocus(this);
    if (!textController.isOpen) {
      textController.openView();
    }
  }

  Widget Function(Iterable<Widget>)? getViewBuilder(BuildContext context) => null;

  List<Widget> buildSuggestions(BuildContext context, SearchController controller) {
    final result = <Widget>[];
    final scope = widget.options.cast().where((e) => e.match(controller.text));
    scope.toList().asMap().forEach((index, e) {
      result.add(ListTile(
        title: itemBuilder(context, e),
        tileColor: index % 2 == 0 ? context.colorScheme.primary.withOpacity(0.05) : null,
        hoverColor: context.colorScheme.primary.withOpacity(0.15),
        onTap: () => onChange(e),
      ));
    });
    return result;
  }

  @override
  Widget buildContent(BuildContext context) {
    final indent = ThemeHelper.getIndent(1.5);
    final hintStyle = context.textTheme.tooltipMedium.copyWith(overflow: TextOverflow.ellipsis);
    final labelStyle = context.textTheme.tooltipSmall;
    K? item = widget.value != null ? widget.options.cast().where((e) => e.equal(widget.value)).firstOrNull : null;
    return SearchAnchor(
      isFullScreen: true,
      searchController: textController,
      viewHintText: widget.tooltip ?? widget.hintText,
      headerHintStyle: hintStyle,
      builder: (context, controller) => TapWidget(
        tooltip: widget.tooltip ?? '',
        onTap: () => onTap(null),
        onFocusChange: (v) => v ? focusController.scrollToFocusedElement(this) : null,
        child: Container(
          width: double.infinity,
          color: context.colorScheme.fieldBackground,
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Expanded(
                child: Padding(
                  padding: EdgeInsets.fromLTRB(
                    indent / 1.5,
                    widget.withLabel && item != null ? 1 : indent,
                    0,
                    widget.withLabel && item != null ? indent / 2 : indent,
                  ),
                  child: item != null
                      ? selectorBuilder(context, item)
                      : Text(
                          widget.hintText ?? '...',
                          style: widget.hintStyle ?? hintStyle,
                        ),
                ),
              ),
              IconButton(
                icon: Icon(Icons.arrow_drop_down, color: widget.hintStyle?.color),
                onPressed: () => controller.openView(),
              ),
            ],
          ),
        ),
      ),
      viewBuilder: getViewBuilder(context),
      suggestionsBuilder: buildSuggestions,
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

![crash-stacktrace](https://github.com/flutter/flutter/assets/8281925/1b40214e-c702-47dc-b2b6-44c41a89145a)
![crash-stacktrace-details](https://github.com/flutter/flutter/assets/8281925/dcf67607-04d5-4677-bec5-4b8b1bbb7c12)

https://github.com/flutter/flutter/assets/8281925/63e6b0f4-78df-4631-8f56-b92ab520317b

https://github.com/flutter/flutter/assets/8281925/66292202-9a80-4a40-838a-f1aabe151b6b

</details>

### Logs

<details><summary>Logs</summary>

```console
Null check operator used on a null value
main.dart.js:70450     at ListSelectorState.onChange$1 (https://lyskouski.github.io/app-finance/main.dart.js:122631:51)
main.dart.js:70450     at ListSelectorState_buildSuggestions__closure.call$0 (https://lyskouski.github.io/app-finance/main.dart.js:122695:25)
main.dart.js:70450     at _InkResponseState.handleTap$0 (https://lyskouski.github.io/app-finance/main.dart.js:155722:14)
main.dart.js:70450     at tear_off.<anonymous> (https://lyskouski.github.io/app-finance/main.dart.js:10803:45)
main.dart.js:70450     at TapGestureRecognizer.invokeCallback$1$3$debugReport (https://lyskouski.github.io/app-finance/main.dart.js:144829:27)
main.dart.js:70450     at TapGestureRecognizer.invokeCallback$2 (https://lyskouski.github.io/app-finance/main.dart.js:144839:19)
main.dart.js:70450     at TapGestureRecognizer.handleTapUp$2$down$up (https://lyskouski.github.io/app-finance/main.dart.js:145178:19)
main.dart.js:70450     at TapGestureRecognizer._checkUp$0 (https://lyskouski.github.io/app-finance/main.dart.js:145118:13)
main.dart.js:70450     at TapGestureRecognizer.acceptGesture$1 (https://lyskouski.github.io/app-finance/main.dart.js:145086:15)
main.dart.js:70450     at GestureArenaManager.sweep$1 (https://lyskouski.github.io/app-finance/main.dart.js:142624:41)

// main.dart.js:70450     at ListSelectorState.onChange$1 (https://lyskouski.github.io/app-finance/main.dart.js:122631:51)
    onChange$1(_, value) {
      var t1, _this = this;
      _this.textController._search_anchor$_anchor._closeView$1(null); // [!ERROR]
      t1 = _this._widget;
      t1.toString;
      t1.setState$1(value.get$id(value));
      t1 = _this.__AbstractSelectorState_focusController_A;
      t1 === $ && A.throwUnnamedLateFieldNI();
      t1.onEditingComplete$1(_this);
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.0, on KDE neon 5.27 6.2.0-37-generic, locale en_US.UTF-8)
    • Flutter version 3.16.0 on channel stable at /home/Programs/flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db7ef5bf9f (7 days ago), 2023-11-15 11:25:44 -0800
    • Engine revision 74d16627b9
    • Dart version 3.2.0
    • DevTools version 2.28.2

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/Android/Sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /home/Programs/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /snap/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.27.5
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2022.3)
    • Android Studio at /home/Programs/android-studio/
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /home/Programs/android-studio/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] VS Code (version 1.84.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.76.0

[✓] Connected device (3 available)
    • SM F926B (mobile) • RFCT10J2GRY • android-arm64  • Android 13 (API 33)
    • Linux (desktop)   • linux       • linux-x64      • KDE neon 5.27 6.2.0-37-generic
    • Chrome (web)      • chrome      • web-javascript • Chromium 119.0.6045.159 snap

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by lyskouski** [comment]
I've resolved the issue by changing:
```dart
final textController = SearchController();
```
to
```
late SearchController textController = SearchController();
```

But it should not be the case for closing the issue. It's totally something wrong in a two-step compilation process (to JS, from JS to JIT)

**Comment by lyskouski** [comment]
up

**Comment by huycozy** [comment]
Hi @lyskouski 
Which platform are you seeing this issue? (Android/iOS) 

Please provide a [completed and minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that doesn’t include 3rd party plugins or complex production code so that we may verify this. Thanks!

**Comment by lyskouski** [comment]
Hello @huycozy,

I've provided a code that leads to the error, with stack trace and descriptions of it reproduction.

It's NOT happened on Web, Android, iOS, macOS, Linux, and Windows.
It's observed on PWA usage (install application from Web) for Android and iOS, since the majority of my users (45%) are using that type of installation.

![image](https://github.com/flutter/flutter/assets/8281925/426b5358-0f12-415b-b46a-becc20a2753c)


It has happened right after the Flutter upgrade from 3.13.9 to 3.16.0.

**Comment by huycozy** [comment]
Can you reproduce the issue locally? It looks like an implementation issue with null-safety. Your sample code is from a complex project, please provide a [completed and minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that doesn’t include 3rd party plugins or complex production code.

**Comment by lyskouski** [comment]
Yes, I've connected phone with installed PWA and debug it via Google Chrome Toolset. It has given me a trace that I've shared in my first post. The minimal reproducible code sample is here: https://github.com/lyskouski/app-finance (as simple as it can be to reproduce the core issue)

The problem is even worse. When the second "compilation" is happening, it ignores the SDK version that was used, alike: CI/CD {[fix] 3.13.9} -> deploy [JS] -> install as PWA [JS -> JIT, 3.16 taken]. Since I've reverted step by step each commit and the error is reproduced even for the initial phase of the project.
